src/mem/cache/prefetch/cdp.hh

```diff
@@ -71,79 +72,329 @@ namespace prefetch
+    {
+      private:
+        std::vector<SubVpnEntry> subEntries;
+        int subEntryNum;
+        int subEntryBits;
+        uint64_t resetPeriod;
+      public:
+        VpnEntry(int num, uint64_t _rstPeriod)
+            : TaggedEntry(),
+              subEntryNum(num),
+              resetPeriod(_rstPeriod)
+        {
+            for (int i = 0; i < subEntryNum; i++) {
+                subEntries.emplace_back(SubVpnEntry(resetPeriod));
             }
+            subEntryBits = ceil(log2(subEntryNum));
         }
-        VpnTable() { resetConfidence(2, false); }
-    } vpnTable;
+        void discard() {
+            for (int i = 0; i < subEntryNum; i++) {
+                subEntries[i].discard();
+            }
+        }
+        void init(uint64_t vpn1, uint64_t vpn2, bool pf_hit_cdp) {
+            uint64_t sub_idx = ((vpn2 << 9) | vpn1) & ((1UL << subEntryBits) - 1);
+            assert(sub_idx < subEntryNum);
+            subEntries[sub_idx].init(vpn1, vpn2, pf_hit_cdp);
+        }
+        void access(uint64_t idx, bool pf_hit_cdp) {
+            subEntries[idx].access(pf_hit_cdp);
+        }
+        void decr(uint64_t idx, bool low_conf) {
+            subEntries[idx].decr(low_conf);
+        }
+        void periodReset(float throttle_aggressiveness, bool enable_thro, bool low_conf) {
+            for (int i = 0; i < subEntryNum; i++) {
+                if (subEntries[i].getExist()) {
+                    subEntries[i].periodReset(throttle_aggressiveness, enable_thro, low_conf);
+                }
+            }
+        }
+        bool getExist(uint64_t idx) { return subEntries[idx].getExist(); }
+        bool getHot(uint64_t idx) { return subEntries[idx].getHot(); }
+        u_int64_t getPrefRefCnt(uint64_t idx) { return subEntries[idx].getPrefRefCnt(); }
+        u_int64_t getDebugVpn1(uint64_t idx) { return subEntries[idx].debug_vpn1; }
+        u_int64_t getDebugVpn2(uint64_t idx) { return subEntries[idx].debug_vpn2; }
+        std::string name() { return std::string("VpnEntry"); }
+    };
+
+    template<class Entry>
+    class VpnTable
+    {
+        private:
+            AssociativeSet<Entry> table;
+            const uint64_t _resetPeriod;
+            uint64_t _resetCounter;
+            uint64_t _subEntryNum;
+            uint64_t _subEntryBits;
+
+        public:
+            VpnTable(int assoc, int num_entries, BaseIndexingPolicy *idx_policy,
+            replacement_policy::Base *rpl_policy, int sub_entries,
+            uint64_t _rst_period, Entry const &init_val = Entry())
+                : table(assoc, num_entries, idx_policy, rpl_policy, init_val),
+                  _resetPeriod(_rst_period),
+                  _resetCounter(0),
+                  _subEntryNum(sub_entries)
+            {
+                assert(_subEntryNum % 2 == 0 && _subEntryNum < 512);
+                _subEntryBits = ceil(log2(_subEntryNum));
+            }
+            void add(int vpn2, int vpn1, bool pf_hit_cdp) {
```

Comment:
Please use the same code format.(235 257)
(1)void aaa() {
}
(2)void aaa()
{
}

Reply:
Thanks I will fix them. orz